core/jvm/src/main/scala/scalaz/zio/RTS.scala

```diff
@@ -1038,12 +1052,14 @@ private object RTS {
     }
 
     private final def purgeJoinersKillers(v: ExitResult[E, A],
+                                          exitHandler: Option[ExitResult[E, A] => IO[Nothing, Unit]],
                                           joiners: List[Callback[E, A]],
                                           killers: List[Callback[E, Unit]]): Unit = {
       // To preserve fair scheduling, we submit all resumptions on the thread
       // pool in (rough) order of their submission.
       killers.reverse.foreach(k => rts.submit(k(SuccessUnit)))
       joiners.foreach(k => rts.submit(k(v)))
+      exitHandler.foreach(k => rts.submit(rts.unsafeRun(k(v))))
```

Comment:
I don't think this is correct, as joiners may wait for something that will only happen in the exit handlers, e.g. in `run`. In which case this is the wrong place for it. What do you think @jdegoes ?

Reply:
that shouldn't matter 🤔  but there's a deadlock somewhere. I'll get to the bottom of it.